WirescaleCli/AccessTokenProvider.cs

```diff
@@ -0,0 +1,125 @@
+            .WithRedirectUrl(_auth0RedirectUri)
+            .WithValue("code_challenge", codeChallenge)
+            .WithValue("code_challenge_method", "S256")
+            .WithScope("openid");
+
+        var authorizationCode = await CaptureAuthorizationCode(authorizationUrlBuilder);
+
+        var accessTokenResponse = await _authenticationApiClient.GetTokenAsync(new AuthorizationCodePkceTokenRequest
+        {
+            Code = authorizationCode,
+            CodeVerifier = codeVerifier,
+            RedirectUri = _auth0RedirectUri,
+            ClientId = _auth0ClientId,
+        });
+
+        return accessTokenResponse.AccessToken;
+    }
+
+    private async Task<String> CaptureAuthorizationCode(AuthorizationUrlBuilder authorizationUrlBuilder)
+    {
+        var callbackHttpListener = new HttpListener();
+        callbackHttpListener.Prefixes.Add($"{_auth0RedirectUri}/");
+
+        callbackHttpListener.Start();
+
+        OpenBrowser(authorizationUrlBuilder.Build());
+
+        var context = await callbackHttpListener.GetContextAsync();
+        var callbackRequest = context.Request;
+
+
+        var authorizationCode = HttpUtility.ParseQueryString(callbackRequest.Url.Query)["code"];
+
+        var response = context.Response;
+        var responseString = "<html><body><p>Authorization code received successfully. You can now close this window.</p></body></html>";
+        var buffer = Encoding.UTF8.GetBytes(responseString);
+
+        response.ContentLength64 = buffer.Length;
+        var responseOutput = response.OutputStream;
+        await responseOutput.WriteAsync(buffer, 0, buffer.Length);
+        responseOutput.Close();
+
+        callbackHttpListener.Stop();
+
+        return authorizationCode;
+    }
+
+    private static void OpenBrowser(Uri uri)
+    {
+        try
+        {
+            Process.Start(new ProcessStartInfo(uri.ToString())
+            {
+                UseShellExecute = true
+            });
+        }
+        catch (Exception exception)
+        {
+            Console.WriteLine($"Failed to open browser: {exception.Message}");
+        }
+    }
+
+    private static string GenerateCodeVerifier()
+    {
+        const int CodeVerifierLength = 64;
+
+        using (var randomNumberGenerator = RandomNumberGenerator.Create())
+        {
+            var bytes = new byte[CodeVerifierLength];
+            randomNumberGenerator.GetBytes(bytes);
+
+            return Base64UrlEncode(bytes);
+        }
+    }
+
+    private static string GenerateCodeChallenge(string codeVerifier)
+    {
+        using (var sha256 = SHA256.Create())
+        {
+            byte[] challengeBytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(codeVerifier));
+
+            return Base64UrlEncode(challengeBytes);
+        }
+    }
+
+    private static string Base64UrlEncode(byte[] data)
+    {
+        var base64 = Convert.ToBase64String(data);
+
+        return base64.Replace("+", "-").Replace("/", "_").Replace("=", "");
```

Comment:
Does C# not have URL-safe base64 encoding in standard library?

Reply:
Ah there was one here https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.tokens.base64urlencoder?view=msal-web-dotnet-latest -- fixed and thanks!